src/builder/context.rs

```diff
@@ -112,26 +117,102 @@ impl<'a> Context<'a> {
         };
     }
 
-    pub fn add_cache_dir(&mut self, path: &Path, name: String)
+    pub fn add_cache_dir(&mut self, path: &Path, name: &str)
         -> Result<(), String>
     {
+        let entry = match self.cache_dirs.entry(path.to_path_buf()) {
+            Entry::Vacant(entry) => entry,
+            Entry::Occupied(_) => return Ok(()),
+        };
```

Comment:
I think it overoptimizes it.
`if self.cache_dirs.contains_key(path) { return ...}`
.. with the insert at the end of the function should be clearer. Also it would avoid `to_path_buf` allocation on `return` path, so might be even better performance-wise.